hudi-spark-datasource/hudi-spark3.0.x/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/Spark30ParquetReader.scala

```diff
@@ -142,11 +149,20 @@ class Spark30ParquetReader(enableVectorizedReader: Boolean,
     }
     val taskContext = Option(TaskContext.get())
     if (enableVectorizedReader) {
-      val vectorizedReader = new VectorizedParquetRecordReader(
-        convertTz.orNull,
-        datetimeRebaseMode.toString,
-        enableOffHeapColumnVector && taskContext.isDefined,
-        capacity)
+      val vectorizedReader = if (schemaEvolutionUtils.shouldUseInternalSchema) {
+        schemaEvolutionUtils.buildVectorizedReader(
+          convertTz,
+          datetimeRebaseMode,
+          enableOffHeapColumnVector,
+          taskContext,
+          capacity)
+      } else {
+        new VectorizedParquetRecordReader(
+          convertTz.orNull,
+          datetimeRebaseMode.toString,
+          enableOffHeapColumnVector && taskContext.isDefined,
+          capacity)
+      }
```

Comment:
Fold the check `schemaEvolutionUtils.shouldUseInternalSchema` into `schemaEvolutionUtils.buildVectorizedReader` (returning new VectorizedParquetRecordReader inside if internal schema is disabled) and simplify the logic here as
```suggestion
      val vectorizedReader = schemaEvolutionUtils.buildVectorizedReader(
          convertTz,
          datetimeRebaseMode,
          enableOffHeapColumnVector,
          taskContext,
          capacity)
```

Reply:
Same for readers for other Spark versions.